src/server/transaction.cc

```diff
@@ -756,8 +760,9 @@ void Transaction::Schedule() {
 // Runs in coordinator thread.
 void Transaction::Execute(RunnableType cb, bool conclude) {
   DCHECK(coordinator_state_ & COORD_SCHED);
+  DCHECK(!cb_ptr_);
 
-  cb_ = std::move(cb);
+  cb_ptr_ = &cb;
```

Comment:
Unless I'm missing something here, this is unsafe. You're taking a ref to a local var.
Perhaps you intended to copy it to a local member of a non-pointer type?
I realize you'd like to know if it's initialized, so you can have the member be of type std::optional<RunnableType>

BTW note that this is somewhat risky, as now callers to this API must make sure that their callable outlive the inner object, but I don't know this API so it might be OK.

Reply:
The lifetime of `cb_/cb_ptr_` is only for during of `ScheduleSingleHop/Execute`.

Reply:
It's true that we could use `std::optional` here but at the end I use `FunctionRef` to accept lambdas which by itself is valid only during the call of these functions.

Reply:
Sure, if the lifetime of `cb_ptr_` is only for the duration of these methods then there's no issue. I didn't realize that, but reading the code now - it makes sense. Sorry about the noise :)

BTW, in the previous version (with `std::function`) the lambdas were moved into the `Transaction` object, so they could have theoretically had a longer lifetime than the function call. I'm not sure if that's what you implied in "... lambdas which by itself is valid only during the call of these functions", but anyway, that's obviously irrelevant.

Reply:
yes, they could - and we paid for it with unnecessary allocation, which I discovered when fixing the bug.
These allocations by themselves are not awful, but their pattern is - to always malloc in one thread and free in another. This puts an unnecessary load on malloc library to move free lists from thread to thread.
those `cb_ = nullptr` are the assignments that reset the function object.